pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -477,13 +516,47 @@ stdenv.mkDerivation ({
     # TODO: fetch the self from the fixpoint instead
     haddockDir = self: if doHaddock then "${docdir self.doc}/html" else null;
 
-    env = shellFor {
-      packages = p: [ drv ];
-      inherit shellHook;
-    };
-
+    envFunc = { withHoogle ? false, ... } @ args:
+      let
+        name = "ghc-shell-for-${drv.name}";
+
+        withPackages = if withHoogle then ghcWithHoogle else ghcWithPackages;
+
+        ghcEnvForBuild = buildHaskellPackages.ghcWithPackages (_: setupHaskellDepends);
+        ghcEnv = withPackages (_:
+          otherBuildInputsHaskell ++
+          propagatedBuildInputs ++
+          stdenv.lib.optionals (!isCross) setupHaskellDepends);
+
+        ghcCommandCaps = stdenv.lib.toUpper ghcCommand';
+
+        mkDrvArgs = builtins.removeAttrs args [ "withHoogle" ];
+      in stdenv.mkDerivation (mkDrvArgs // {
+        name = mkDrvArgs.name or name;
+
+        depsBuildBuild = stdenv.lib.optional isCross ghcEnvForBuild;
+        nativeBuildInputs =
+          [ ghcEnv ] ++ optional (allPkgconfigDepends != []) pkgconfig ++
+          collectedToolDepends ++
+          mkDrvArgs.nativeBuildInputs or [];
+        buildInputs =
+          otherBuildInputsSystem ++
+          mkDrvArgs.buildInputs or [];
+        phases = ["installPhase"];
+        installPhase = "echo $nativeBuildInputs $buildInputs > $out";
+        LANG = "en_US.UTF-8";
+        LOCALE_ARCHIVE = stdenv.lib.optionalString (stdenv.hostPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
```

Comment:
As far as I understand, if we're cross compiling, this shell derivation is meant to be run on the host system.

In that case, doesn't it make more sense to pull `glibcLocales` from normal `pkgs` instead of `buildPackages`?

Reply:
The shell is run on the build platform, but `ghcEnv` will produce binaries for the host platform.

Reply:
I thought about this a little bit more, and I realized my mistake here.

The shell created by `shellFor` or `env` is the shell that contains the cross-compiling GHC.  So, for instance, it would run on x86_64 but compile for amd.

It makes sense that the `glibcLocales` need to be pulled from `buildPackages`.